pricegraph/src/orderbook.rs

```diff
@@ -129,15 +130,80 @@ impl Orderbook {
 
     /// Reduces the orderbook by matching all overlapping ring trades.
     pub fn reduce_overlapping_orders(&mut self) {
+        Subgraphs::new(self.projection.node_indices())
+            .for_each(|token| self.reduced_shortest_paths(token));
+    }
+
+    /// Fill a market order in the current orderbook graph returning the maximum
+    /// price the order can have while overlapping with existing orders. Returns
+    /// `None` if the order cannot be filled because the token pair is not
+    /// connected, or the connection cannot support enough capacity to fully
+    /// fill the specified volume.
+    ///
+    /// This is done by pushing the specified volume as a flow through the graph
+    /// and finding the min-cost paths that have enough capacity to carry the
+    /// flow from the sell token to the buy token. The min-cost is calculated
+    /// using the successive shortest path algorithm with Bellman-Ford.
+    ///
+    /// Note that this method is not idempotent, successive calls with the same
+    /// token pair and volume yield different results as orders get filled with
+    /// each match changing the orderbook.
+    pub fn fill_market_order(&mut self, pair: TokenPair, volume: f64) -> Option<f64> {
         self.update_projection_graph();
 
-        Subgraphs::new(self.projection.node_indices()).for_each(|token| loop {
-            let mut negative_cycle_predecessor = None;
-            match bellman_ford::search(&self.projection, token) {
-                Ok((_, predecessor)) => break negative_cycle_predecessor.unwrap_or(predecessor),
-                Err(NegativeCycle(predecessor, start)) => {
+        let (sell, buy) = (node_index(pair.sell), node_index(pair.buy));
```

Comment:
Maybe my logic is wrong but doesn't this have to be inversed? If you are trying to fill a market order selling ETH for DAI, wouldn't you be looking in the graph for paths that are willing to buy ETH for DAI and check their price?

Reply:
That is correct, it is why we are searching from the sell token since orders produce graph edges that go from the buy token to the sell token.